RecoHGCal/TICL/plugins/EGammaSuperclusterProducer.cc

```diff
@@ -0,0 +1,124 @@
+
+private:
+  edm::EDGetTokenT<ticl::TracksterCollection> ticlSuperClustersToken_;
+  edm::EDGetTokenT<std::vector<std::vector<unsigned int>>> superClusterLinksToken_;
+  edm::EDGetTokenT<ticl::TracksterCollection> ticlTrackstersEMToken_;
+  edm::EDGetTokenT<reco::CaloClusterCollection> layerClustersToken_;
+};
+
+EGammaSuperclusterProducer::EGammaSuperclusterProducer(const edm::ParameterSet& ps)
+    : ticlSuperClustersToken_(consumes<ticl::TracksterCollection>(ps.getParameter<edm::InputTag>("ticlSuperClusters"))),
+      superClusterLinksToken_(consumes<std::vector<std::vector<unsigned int>>>(
+          edm::InputTag(ps.getParameter<edm::InputTag>("ticlSuperClusters").label(),
+                        "linkedTracksterIdToInputTracksterId",
+                        ps.getParameter<edm::InputTag>("ticlSuperClusters").process()))),
+      ticlTrackstersEMToken_(consumes<ticl::TracksterCollection>(ps.getParameter<edm::InputTag>("ticlTrackstersEM"))),
+      layerClustersToken_(consumes<reco::CaloClusterCollection>(ps.getParameter<edm::InputTag>("layerClusters"))) {
+  produces<reco::SuperClusterCollection>();
+  produces<reco::CaloClusterCollection>();  // The CaloCluster corresponding to each EM trackster
+}
+
+void EGammaSuperclusterProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  auto const& ticlSuperclusters = iEvent.get(ticlSuperClustersToken_);
+  auto const& ticlSuperclusterLinks = iEvent.get(superClusterLinksToken_);
+  auto emTracksters_h = iEvent.getHandle(ticlTrackstersEMToken_);
+  auto const& emTracksters = *emTracksters_h;
+  auto const& layerClusters = iEvent.get(layerClustersToken_);
+  // Output collections :
+  auto egammaSuperclusters = std::make_unique<reco::SuperClusterCollection>();
+  auto caloClustersEM = std::make_unique<reco::CaloClusterCollection>();
+
+  // Fill reco::CaloCluster collection (1-1 mapping to TICL EM tracksters)
+  for (ticl::Trackster const& emTrackster : emTracksters) {
+    std::vector<std::pair<DetId, float>> hitsAndFractions;
+    int iLC = 0;
+    std::for_each(std::begin(emTrackster.vertices()), std::end(emTrackster.vertices()), [&](unsigned int lcId) {
+      const auto fraction = 1.f / emTrackster.vertex_multiplicity(iLC++);
+      for (const auto& cell : layerClusters[lcId].hitsAndFractions()) {
+        hitsAndFractions.emplace_back(cell.first, cell.second * fraction);
+      }
+    });
+
+    reco::CaloCluster& caloCluster = caloClustersEM->emplace_back(
+        emTrackster.regressed_energy(),            // energy
+        math::XYZPoint(emTrackster.barycenter()),  // position
+        reco::CaloID(reco::CaloID::DET_HGCAL_ENDCAP),
+        hitsAndFractions,
+        reco::CaloCluster::particleFlow,  // algoID (copying from output of PFECALCSuperClusterProducer)
+        hitsAndFractions.at(0).first  // seedId : TODO figure out if needed or if we need to put the highest energy hit
+    );
+    caloCluster.setCorrectedEnergy(emTrackster.regressed_energy());
```

Comment:
Is it expected to have the regressed energy for the CLUE3D Tracksters? Is the plan to have an energy regression model on the CLUE3D EM Tracksters before the SuperClustering?

Reply:
This still needs to be determined. We need to figure out if the CaloCluster corrected energy is used anywhere in the GsfTrack chain.